components/tidb_query/src/codec/row/v2/decoder.rs

```diff
@@ -0,0 +1,332 @@
+use crate::codec::{
+    mysql::{self, DecimalDecoder, JsonDecoder, Time},
+    Datum, Result,
+};
+use crate::expr::EvalContext;
+use codec::number::NumberCodec;
+use codec::prelude::*;
+use tidb_query_datatype::FieldTypeAccessor;
+use tikv_util::collections::HashMap;
+use tipb::ColumnInfo;
+
+pub trait RowDecoder: NumberDecoder + IntDecoder {
+    // | version | flag | number_of_not_null_values | number_of_null_values | not_null_value_ids | null_value_ids | value_offsets | values
+    fn decode_row(
+        &mut self,
+        ctx: &EvalContext,
+        cols: &HashMap<i64, ColumnInfo>,
+    ) -> Result<HashMap<i64, Datum>> {
+        let mut row = HashMap::with_capacity_and_hasher(cols.len(), Default::default());
```

Comment:
Any benefit to use `HashMap::with_capacity_and_hasher(cols.len(), Default::default());`?